contentcuration/contentcuration/utils/pagination.py

```diff
@@ -1,84 +1,121 @@
 import hashlib
 
 from django.core.cache import cache
+from django.core.exceptions import EmptyResultSet
 from django.core.paginator import InvalidPage
+from django.core.paginator import Page
+from django.core.paginator import Paginator
+from django.db.models import QuerySet
+from django.utils.functional import cached_property
 from rest_framework.pagination import NotFound
 from rest_framework.pagination import PageNumberPagination
 from rest_framework.response import Response
 
 
-class CachedListPagination(PageNumberPagination):
-    def paginate_queryset(self, queryset, request, view=None):
-        """
-        Overrides paginate_queryset from PageNumberPagination
-        to assign the records count to the paginator.
-        This count has been previously obtained in a less complex query,
-        thus it has a better performance.
+class ValuesPage(Page):
+    def __init__(self, object_list, number, paginator):
+        self.queryset = object_list
+        self.object_list = object_list
+        self.number = number
+        self.paginator = paginator
 
-        """
 
-        def cached_count(queryset):
+class ValuesViewsetPaginator(Paginator):
+    def __init__(self, object_list, *args, **kwargs):
+        if not isinstance(object_list, QuerySet):
+            raise TypeError("ValuesViewsetPaginator is only intended for use with Querysets")
+        self.queryset = object_list
+        object_list = object_list.values_list("pk", flat=True).distinct()
+        return super(ValuesViewsetPaginator, self).__init__(object_list, *args, **kwargs)
+
+    def _get_page(self, object_list, *args, **kwargs):
+        pks = list(object_list)
+        return ValuesPage(self.queryset.filter(pk__in=pks), *args, **kwargs)
+
+
+class CachedValuesViewsetPaginator(ValuesViewsetPaginator):
+    @cached_property
```

Comment:
I understand the double caching here through `cached_property` and the internal caching, but it took me second, especially since the previous pagination functionality instantiated the pagination class only once.

Reply:
Can add a comment.